test/e2e/antreapolicy_test.go

```diff
@@ -1039,30 +1093,65 @@ func testACNPClusterGroupRefRulePodAdd(t *testing.T, data *TestData) {
 	executeTestsWithData(t, testCase, data)
 }
 
+// There are three situations of a Pod's IP(s):
+// 1. Only one IPv4 address.
+// 2. Only one IPv6 address.
+// 3. One IPv4 and one IPv6 address, and we don't know the order in list.
+// We need to add all IP(s) of Pods as CIDR to IPBlock.
+func genIPBlock(ip string) *crdv1beta1.IPBlock {
+	switch IPFamily(ip) {
+	case "v4":
+		return &crdv1beta1.IPBlock{CIDR: ip + "/32"}
+	case "v6":
+		return &crdv1beta1.IPBlock{CIDR: ip + "/128"}
+	default:
+		return nil
+	}
+}
+
+// genIPBlockWithExceptIPs generates ipBlocks which contains all the IP addresses in the
+// provided IPs' address family(s), except for the addresses in the input slice.
+func genIPBlockWithExceptIPs(except []string) []*crdv1beta1.IPBlock {
+	var v4Excepts, v6Excepts []string
+	var ipbs []*crdv1beta1.IPBlock
+	for _, e := range except {
+		if IPFamily(e) == "v4" {
+			v4Excepts = append(v4Excepts, e+"/32")
+		} else if IPFamily(e) == "v6" {
+			v6Excepts = append(v6Excepts, e+"/128")
+		}
+	}
+	if len(v4Excepts) > 0 {
+		ipbs = append(ipbs, &crdv1beta1.IPBlock{
+			CIDR:   anyIPv4,
+			Except: v4Excepts,
+		})
+	}
+	if len(v6Excepts) > 0 {
+		ipbs = append(ipbs, &crdv1beta1.IPBlock{
+			CIDR:   anyIPv6,
+			Except: v6Excepts,
+		})
+	}
+	return ipbs
+}
+
 func testACNPClusterGroupRefRuleIPBlocks(t *testing.T) {
 	podXAIP, _ := podIPs[getPodName("x", "a")]
 	podXBIP, _ := podIPs[getPodName("x", "b")]
 	podXCIP, _ := podIPs[getPodName("x", "c")]
 	podZAIP, _ := podIPs[getPodName("z", "a")]
-	// There are three situations of a Pod's IP(s):
-	// 1. Only one IPv4 address.
-	// 2. Only one IPv6 address.
-	// 3. One IPv4 and one IPv6 address, and we don't know the order in list.
-	// We need to add all IP(s) of Pods as CIDR to IPBlock.
-	genCIDR := func(ip string) string {
-		if strings.Contains(ip, ".") {
-			return ip + "/32"
-		}
-		return ip + "/128"
-	}
 	var ipBlock1, ipBlock2 []crdv1beta1.IPBlock
 	for i := 0; i < len(podXAIP); i++ {
-		ipBlock1 = append(ipBlock1, crdv1beta1.IPBlock{CIDR: genCIDR(podXAIP[i])})
-		ipBlock1 = append(ipBlock1, crdv1beta1.IPBlock{CIDR: genCIDR(podXBIP[i])})
-		ipBlock1 = append(ipBlock1, crdv1beta1.IPBlock{CIDR: genCIDR(podXCIP[i])})
-		ipBlock2 = append(ipBlock2, crdv1beta1.IPBlock{CIDR: genCIDR(podZAIP[i])})
+		for _, ips := range [][]string{podXAIP, podXBIP, podXCIP} {
+			if ipb := genIPBlock(ips[i]); ipb != nil {
```

Comment:
I feel like the fact that `genIPBlock` can return `nil` is just making the test code more verbose

Reply:
Refactored